lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,204 @@
+//===------ YieldOnceCheck.cpp - Check usage of yields in accessors  ------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "yield-once-check"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "llvm/ADT/DenseSet.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/AST/DiagnosticsSIL.h"
+
+using namespace swift;
+
+namespace {
+
+class YieldOnceCheck : public SILFunctionTransform {
+
+  template<typename...T, typename...U>
+  static InFlightDiagnostic
+  diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag, U &&...args) {
+    return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
```

Comment:
You should consider pushing this up to `SILTransform`, but it's not really harmful to declare it here if that's common practice for other diagnostic passes.